docs/extensions.md

```diff
@@ -0,0 +1,354 @@
+| Shown in Library               | X    | √    | √        | X          |
+| Sandboxed                      | X    | X    | √        | √          |
+| Can save projects to community | √    | √    | √        | X          |
+
+## JavaScript Environment
+
+Most Scratch 3.0 is written using JavaScript features not yet commonly supported by browsers. For compatibility we
+transpile the code to ES5 before publishing or deploying. Any extension included in the `scratch-vm` repository may
+use ES6+ features and may use `require` to reference other code within the `scratch-vm` repository.
+
+Unofficial extensions must be self-contained. Authors of unofficial extensions are responsible for ensuring browser
+compatibility for those extensions, including transpiling if necessary.
+
+## Translation
+
+Scratch extensions use the [ICU message format](http://userguide.icu-project.org/formatparse/messages) to handle
+translation across languages. For **core, team, and official** extensions, the function `formatMessage` is used to
+wrap any ICU messages that need to be exported to the [Scratch Transifex group](https://www.transifex.com/llk/public/)
+for translation.
+
+**All extensions** may additionally define a `translation_map` object within the `getInfo` function which can provide
+translations within an extension itself. The "Annotated Example" below provides a more complete illustration of how
+translation within an extension can be managed. **WARNING:** the `translation_map` feature is currently in the
+proposal phase and may change before implementation.
+
+## Backwards Compatibility
+
+Scratch is designed to be fully backwards compatible. Because of this, block definitions and opcodes should *never*
+change in a way that could cause previously saved projects to fail to load or to act in unexpected / inconsistent
+ways.
+
+## Defining an Extension
+
+Scratch extensions are defined as a single Javascript class which accepts either a reference to the Scratch
+[VM](https://github.com/llk/scratch-vm) runtime or a "runtime proxy" which handles communication with the Scratch VM
+across a well defined worker boundary (i.e. the sandbox).
+
+```js
+class SomeBlocks {
+    constructor (runtime) {
+        /**
+         * Store this for later communication with the Scratch VM runtime.
+         * If this extension is running in a sandbox then `runtime` is an async proxy object.
+         * @type {Runtime}
+         */
+        this.runtime = runtime;
+    }
+
+    // ...
+}
+```
+
+All extensions must define a function called `getInfo` which returns an object that contains the information needed to
+render both the blocks and the extension itself.
+
+```js
+class SomeBlocks {
+    // ...
+    getInfo () {
+        return {
+            id: 'someBlocks',
+            name: 'Some Blocks',
+            blocks: [
+                {
+                    opcode: 'myReporter',
+                    blockType: BlockType.REPORTER,
```

Comment:
Should we include the `require`s for both `BlockType` and `ArgumentType` in this example?